[LoopPeel] Add new option to peeling loops to convert PHI into IV

kasuga-fj opened the pull request:
LoopPeel currently considers PHI nodes that become loop invariants through peeling. However, in some cases, peeling transforms PHI nodes into induction variables (IVs), potentially enabling further optimizations such as loop vectorization. For example:

```c
// TSVC s292
int im = N-1;
for (int i=0; i<N; i++) {
  a[i] = b[i] + b[im];
  im = i;
}
```

In this case, peeling one iteration converts `im` into an IV, allowing it to be handled by the loop vectorizer.

This patch adds a new feature to peel loops when to convert PHIs into IVs. At the moment this feature is disabled by default.

Enabling it allows to vectorize the above example. I have measured on neoverse-v2 and observed a speedup of more than 60% (options: `-O3 -ffast-math -mcpu=neoverse-v2 -mllvm -enable-peeling-for-iv`).

This PR is taken over from #94900
Related #81851

kasuga-fj commented:
ping

madhur13490 commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -175,13 +202,17 @@ class PhiAnalyzer {
+  PeelCounter calculate(Value &, ScalarEvolution &SE);
+
+  // Returns true if the \p Phi is an induction in the target loop. This
+  // funciton is a wrapper of `InductionDescriptor::isInductionPHI`.
```
Typo fuciton -> function.

kasuga-fj replied:
Fixed, thanks.

kasuga-fj commented:
@nikic Could you please evaluate the compile-time difference of this PR?

kasuga-fj commented:
> 2. Can you please add a test case for the below case too?

Looks like I added the wrong test case, I will fix it. Thanks.

nikic commented:
Compile-time: https://llvm-compile-time-tracker.com/compare.php?from=8fb29ba287d72392bd7900c33d2a8d2149126dbe&to=060d43fb0d920a663d5199ff37ebfdf204009177&stat=instructions:u

Haven't looked at the implementation yet...

kasuga-fj commented:
Thanks!

@madhur13490 What do you think about the impact? I'm not sure if this is large regression or not. The +0.52% for SPASS is not negligible?

madhur13490 commented:
> @madhur13490 What do you think about the impact? I'm not sure if this is large regression or not. The +0.52% for SPASS is not negligible?

I am fine as long as @nikic is!

madhur13490 commented:
Hi @jamieschmeiser @fhahn can you please also have a look?

jamieschmeiser commented:
My only comments are stylistic.  The reason I added the PhiAnalyzer and brought the code into the class when I reworked/expanded the algorithms was that it organized things and cleaned up the calling between the various functions.  In that spirit, I would suggest adding ScalarEvolution as a member to PhiAnalyzer and setting it in the constructor rather than passing it around.  This is a minor stylistic nit but (I think) it makes the code cleaner and more readable.  I certainly wouldn't block this change for that if you do not agree.  Thanks for doing this.

kasuga-fj commented:
Hi @jamieschmeiser, I don't have a strong opinion on the style and have decided to follow the original thought. Thanks for your comment.

madhur13490 commented:
LGTM. Please wait for @nikic for confirmation on compile-time impact.

kasuga-fj commented:
I found that the SCEV calculation to call `isInductionPHI` was expensive. I replaced it with cheaper checks and the compilation time regression was mitigated on my local.

@nikic Could you evaluate the compile-time again?

kasuga-fj committed:
[LoopPeel] Peel to make Phis loop inductions

LoopPeel now only handles Phis when they become loop invariants by
peeling. There are cases where peeling makes Phis loop invariants, and
peeling in such cases is also useful for other optimizations, such as
loop vectorization. For example, consider the following loops.

```
int im = N-1;
for (int i=0;i<N;i++) {
  a[i] = b[i]+b[im];
  im = i;
}
```

In this case, peeling by 1 iteration makes `im` a loop induction, so we
can vectorize the loop.
This patch allows to vectorize the kernel of s291 and s292 in TSVC. I
have measured on neoverse-v2 and  observed a speedup of more than 60%
(options: `-O3 -ffast-math -mcpu=neoverse-v2`).
Note that in some cases there was unnecessary peeling when tried with
llvm-test-suite. The causes include peeling for a remainder loop of
vectorization and the limitations of analysis by SCEV. However, as far
as I've tried, these unnecessary peels do not affect performance.

kasuga-fj commented:
Gentle ping

madhur13490 commented:
@kasuga-fj I don't think compile-time measurements *has* to be done by Nikita only. You can also follow the steps on compile-time-tracker and get it triggered automatically. If your repo is already on compile-time-tracker's radar, following the branch naming convention will trigger the run automatically. If you have already passed that stage, then please post your results.

nikic commented:
Ah sorry, I did test this back then but forgot the report the results. They look good: https://llvm-compile-time-tracker.com/compare.php?from=8374d421861cd3d47e21ae7889ba0b4c498e8d85&to=53ef622e8ef3d26660f29c6dff4d1fb869270a69&stat=instructions:u

kasuga-fj committed:
Apply suggestions from code review

Co-authored-by: Nikita Popov <github@npopov.com>

kasuga-fj committed:
Apply clang-format

kasuga-fj committed:
Address review comments

kasuga-fj commented on llvm/test/Transforms/LoopUnroll/peel-loop-phi-analysis.ll:
Added two cases where the peeling is not desirable. The current implementation only considers constant increments, so I don't think there are any cases where peeling does not result in IV (there are cases where the variable is originally IV without peeling).

kasuga-fj commented on llvm/test/Transforms/LoopUnroll/peel-loop-phi-analysis.ll:
```diff
@@ -197,3 +197,223 @@ for.body:
+;     a[i] = b[i] + b[im];
+;     im = i;
+;   }
+; }
```
Fixed

kasuga-fj reviewed (commented):
> > Note that in some cases there was unnecessary peeling when tried with llvm-test-suite. The causes include peeling for a remainder loop of vectorization and the limitations of analysis by SCEV. However, as far as I've tried, these unnecessary peels do not affect performance.
>
> Do you have an example handy? Generally having some unnecessary peeling is expected, but peeling remainder loops is not great.

I found such cases when I first submitted this patch (I used SCEV to detect IVs at the time), but I just tried to run llvm-test-suite and it didn't reproduce. I'm now checking the details of the difference with and without this patch, but there does not appear to be a significant difference. The following loops were newly peeled.

EDIT: Some peelings in the following list is unnecessary one.  They are fixed in https://github.com/llvm/llvm-project/pull/121104/commits/c6f1a3a074dfbeb56bb1f824c7608853333878b and the peeling no longer happens. I've checked that all others are what peeled as intended.

- `MultiSource/Applications/ClamAV/libclamav_matcher-ac.c:888:6`
- ~`MultiSource/Applications/JM/ldecod/vlc.c:886:3`~
- ~`MultiSource/Applications/JM/ldecod/vlc.c:961:3`~
- ~`MultiSource/Applications/JM/lencod/cabac.c:1492:3`~
- ~`MultiSource/Applications/oggenc/oggenc.c:47108:7`~
- `MultiSource/Benchmarks/Prolangs-C/TimberWolfMC/finalpin.c:434:2`
- `MultiSource/Benchmarks/TSVC/CrossingThresholds-dbl/../tsc.inc:3310:3`
- `MultiSource/Benchmarks/TSVC/CrossingThresholds-dbl/../tsc.inc:3342:3`
- `MultiSource/Benchmarks/TSVC/CrossingThresholds-flt/../tsc.inc:3310:3`
- `MultiSource/Benchmarks/TSVC/CrossingThresholds-flt/../tsc.inc:3342:3`

kasuga-fj committed:
Move the comments of the test outisde the function

kasuga-fj committed:
Fix the handling of binary instructions for inductions

kasuga-fj committed:
Apply clang-format

kasuga-fj commented:
Updated the PR description.

kasuga-fj commented:
Ping

kasuga-fj committed:
Fix comment

kasuga-fj committed:
Remove nsw/nuw check

kasuga-fj committed:
Add an option to toggle peeling to make PHIs into IVs

kasuga-fj committed:
Revert unnecessary change

kasuga-fj commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -190,6 +221,67 @@ PhiAnalyzer::PhiAnalyzer(const Loop &L, unsigned MaxIterations)
+      if (BinOp->getOpcode() != Instruction::Add &&
+          BinOp->getOpcode() != Instruction::Sub)
+        return false;
+      if (!BinOp->hasNoUnsignedWrap() || !BinOp->hasNoSignedWrap())
```
Now I don't think this check is necessary. However, removing them happened undesirable peelings. It seems that this check just happens to work well to deflect undesired loops.

kasuga-fj reviewed (commented):
Currently there are some cases where undesirable peeling happens. As I checked, there were roughly two major patterns:

- Failed to analysis cast operations, like `induction_undesirable_peel1` I added in test.
- Peeling is performed properly to make PHIs into IVs, but such a peeling doesn't seem to be profitable. 
  - The main purpose of such peelings are for vectorization, but the loop cannot be vectorized due to other reason, e.g., it has function calls.

I think it's reasonable to address these issues in other patches. So, for now, I want to control this feature with a flag. After they are resolved, I'd like to enable it by default.

kasuga-fj commented:
Gentle ping @nikic , thanks!

kasuga-fj commented:
Gentle ping @nikic . 
For now, I've added a flag and disabled this feature by default, so I don't think this patch will affect existing cases. I'd like to enable it by default in the future, after addressing the cases in `peel-loop-phi-analysis-iv.ll`, where undesired peeling happens.

madhur13490 commented:
@kasuga-fj Are you planning to proceed on this patch?

kasuga-fj committed:
Merge branch 'main' into peeling-loop-induction

kasuga-fj committed:
Revise comments

kasuga-fj committed:
Revise tests

kasuga-fj committed:
Remove unrelated change

madhur13490 commented:
Can we please progress on this PR? We are tracking this as it addresses one of the tests we are currently examining.

kasuga-fj commented:
Okay. Since this has already been approved, I'll wait for a week, and if there are no further objections, I'll go ahead and merge it.

kasuga-fj committed:
Merge branch 'main' into peeling-loop-induction

kasuga-fj commented:
Since there don't seem to be any objections, I'd like to go ahead and merge this.

kasuga-fj committed:
Merge branch 'main' into peeling-loop-induction

nikic commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -155,45 +159,170 @@ namespace {
 
+  // Auxiliary function to calculate the number of iterations for a comparison
+  // instruction or a binary operator.
+  PeelCounter mergeTwoCounter(const Instruction &CmpOrBinaryOp,
```
nit: mergeTwoCounters?

nikic commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -155,45 +159,170 @@ namespace {
+      break;
+
+    // Avoid infinite loop.
+    if (Visited.contains(Cur))
```
Use `if (!Visited.insert(Cur))` to avoid duplicate hash lookup.

nikic commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -155,45 +159,170 @@ namespace {
 
-  // Calculate the number of iterations after which the given value
-  // becomes an invariant.
+  // Return a value representing zero for the given counter type.
```
Somewhat a pre-existing issue, but these comments should use `///`.

nikic commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -155,45 +159,170 @@ namespace {
+
+/// When either \p LHS or \p RHS is an IV, the result of \p CmpOrBinaryOp is
+/// considered an IV only if it is an addition or a subtraction. Otherwise the
+/// result can be a value that is neither an loop-invariant nor an IV.
```
an -> a

nikic reviewed (commented):
Better late than never: This LGTM.
